Custom Document transforms docs (#11089)
apollographql/apollo-client: docs/source/data/document-transforms.mdx

Author: Something I'm realizing that really needs a tweak is that I only show one code samples that uses `ApolloClient` at all. I need to update more of the code samples to show the whole flow (i.e. when network calls are made). This focuses too much on the `.transformDocument()` API which users won't typically use directly.

---
title: Document transforms
description: Make custom modifications to your GraphQL documents
---

import DocumentTransformOptions from '../../shared/document-transform-options.mdx';

> > This article assumes you're familiar with the [anatomy of a GraphQL query](https://www.apollographql.com/blog/graphql/basics/the-anatomy-of-a-graphql-query/) and the concept of an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree). To explore a GraphQL AST, visit [AST Explorer](https://astexplorer.net/).

Reviewer: Do you know of any documentation for the graphql AST specifications? I actually struggled with this when working on another project using the AST, and the only way I was able to decipher the different nodes was through the AST explorer and console logging nodes through the `visit` function (which was a bit tedious).  If that’s available, it might be a good resource for this documentation!

Author: I tried looking for it, but it was super difficult to find it documented anywhere. I'll keep searching to see if I can link to a better location, but great callout!

> > This article assumes you're familiar with the [anatomy of a GraphQL query](https://www.apollographql.com/blog/graphql/basics/the-anatomy-of-a-graphql-query/) and the concept of an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree). To explore a GraphQL AST, visit [AST Explorer](https://astexplorer.net/).

Author: This is definitely one of those gray areas in my mind because I could go either way. If you don't already know what an AST is, this feature might be pretty difficult for you to use. But... I don't know that its a strong enough reason to avoid this, so I think its fine to spell out the acronym. Thanks for the suggestion!

> Have you noticed that Apollo Client makes modifications to your queries, such as adding the `__typename` field, before those queries are sent to your GraphQL server? Apollo Client provides an advanced capability that allows you to define your own GraphQL document transforms that make modifications to your GraphQL queries. Let's learn how you can make and utilize your own GraphQL document transforms.

Author: Thank you so much! This was one of my least favorite sentences, so I was hoping you'd come in and provide some guidance here. I might adapt this slightly, but otherwise this is much better than what I had. Thanks!

> This feature requires Apollo Client `v3.8` or greater.

## Overview

> Document transforms make modifications to GraphQL documents. A GraphQL document is an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) that defines one or more operations and fragments, parsed from a raw GraphQL query string using the `gql` function. You can create your own document transforms using the `DocumentTransform` class. The created transform is then passed to the `ApolloClient` constructor.

Author: Ah yes, the old active vs passive voice 😆. I went through a lesson on this subject during my time at New Relic, but I'm afraid I forgot how to remember the difference 😭. Thanks for calling this out!

> Document transforms make modifications to GraphQL documents. A GraphQL document is an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) that defines one or more operations and fragments, parsed from a raw GraphQL query string using the `gql` function. You can create your own document transforms using the `DocumentTransform` class. The created transform is then passed to the `ApolloClient` constructor.

Reviewer 2: And whoops, meant to share this pretty good resource: https://www.grammarly.com/blog/active-vs-passive-voice/

> Document transforms make modifications to GraphQL documents. A GraphQL document is an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) that defines one or more operations and fragments, parsed from a raw GraphQL query string using the `gql` function. You can create your own document transforms using the `DocumentTransform` class. The created transform is then passed to the `ApolloClient` constructor.

Author: Thanks! I tweaked this just a little more and landed with [078bf70](https://github.com/apollographql/apollo-client/pull/11089/commits/078bf707c1206fdab46f8f3814051d52350e419a).

```ts
import { DocumentTransform } from '@apollo/client';

const documentTransform = new DocumentTransform((document) => {
  // modify the document
  return transformedDocument;
});

const client = new ApolloClient({
  documentTransform
});
```

### Lifecycle 

Document transforms are run before every GraphQL request, for all operations, made through Apollo Client. This extends to any API that performs a network request, such as the [`useQuery`](/react/api/react/hooks#usequery) hook or the [`refetch`](/react/api/core/ObservableQuery#refetch) function on [`ObservableQuery`](/react/api/core/ObservableQuery).

> Document transforms are run very early in the request's lifecycle which makes it possible for modifications to GraphQL documents to be seen by the cache; an important distinction from modifications made to GraphQL documents in an [Apollo Link](/react/api/link/introduction). Because document transforms are run very early in the request lifecycle, it is also possible to add [`@client`](/react/data/directives#client) directives to fields in your document transform to turn the field into a [local-only field](/react/local-state/managing-state-with-field-policies).

Author: Thanks! I tweaked this section a bit more between your and Maria's suggestion and went with [4491727](https://github.com/apollographql/apollo-client/pull/11089/commits/44917270b8da6400dce40ef14c0d4ed7d6f93182)

```ts {2-4}
const transformedDocument = visit(document, {
  Field(fieldNode) {
    // ...
  }
});
```

> In this example, we are using the shorthand visitor syntax, which defines the `enter` function on this node for us. This is equivalent to the following:
> ```ts
> visit(document, {
>   Field: {
>     enter(fieldNode) {
>       // ...
>     }
>   }
> });
> ```

Our document transform only needs to make modifications to a field named `currentUser`, so we need to check the field node's `name` to know if we are working with the `currentUser` field. Let's add a conditional check and return early if we encounter any field not named `currentUser`.

```ts
const transformedDocument = visit(document, {
  Field(fieldNode) {
    if (fieldNode.name.value !== 'currentUser') {
      return;
    }
  }
});
```
> Returning `undefined` from our `Field` visitor tells the `visit` function to leave the node unchanged.

Now that we've determined we are working with the `currentUser` field, we need to figure out if our `id` field is already part of the `currentUser` field's selection set. By checking it's selection set, we ensure we don't accidentally select the field twice in our query. 

> To do so, let's get the field's `selectionSet` and loop over it's `selections` to determine if it includes a field named `id`. It's important to note that a `selectionSet` may contain `selections` of both fields and fragments. Our implementation will need to perform a check against the selection's `kind` property to limit our check only to fields and not fragments. If we find a match on a field named `id`, we no longer need to do any work and can stop traversal of the AST. 

Author: Yes! This still makes sense without that first part. Will apply this change!

        selectionNode.name.value === 'id'
      ) {
        return BREAK;
      }
    }
  }
});
```

> To keep our document transform simple, it does not traverse fragments within the `currentUser` field to determine if those fragments contain an `id` field. A more complete version of this document transform might perform this check.

Now that we know the `id` field is missing, we can add it to our `currentUser` field's selection set. To do so, let's first create a new field node and give it a name of `id`. This is just a plain object with the `kind` property set to `Kind.FIELD` and a `name` node that defines the field's name.

```ts
const idField = {
  kind: Kind.FIELD,
  name: {
    kind: Kind.NAME,
    value: 'id',
  },
};
```

We can now return a modified field node from our visitor that adds that field to the node's `selectionSet` to update our GraphQL document.

```ts {8-14}
const transformedDocument = visit(document, {
  Field(fieldNode) {
    // ...
    const idField = {
      // ... 
    };

    return {
      ...fieldNode,
      selectionSet: {
        ...fieldNode.selectionSet,
        selections: [...selections, idField],
      },
    };
  }
});
```

> > Here we've added the `id` field to the end of the selection set. The order does not matter and you may prefer to put the field in a different order within the `selections` array.

Reviewer 2: ```suggestion
> This example adds the `id` field to the end of the selection set. Order doesn't matter—you may prefer to put the field elsewhere in the `selections` array.
```

> Hooray! We've now got a working document transform that ensures the `id` field is selected whenever a query is sent to our server that contains the `currentUser` field. For completeness, here is the full definition of our custom document transform after completing this example.

Reviewer 2: ```suggestion
Hooray! We now have a working document transform that ensures the `id` field is selected whenever a query containing the `currentUser` field is sent to our server. For completeness, here is the full definition of our custom document transform after completing this example.
```

        if (
          selectionNode.kind === Kind.FIELD &&
          selectionNode.name.value === 'id'
        ) {
          return BREAK;
        }
      }

      const idField = {
        kind: Kind.FIELD,
        name: {
          kind: Kind.NAME,
          value: 'id',
        },
      };

      return {
        ...fieldNode,
        selectionSet: {
          ...fieldNode.selectionSet,
          selections: [...selections, idField],
        },
      };
    },
  });

  return transformedDocument;
});
```

### Checking our document transform

We can check our custom document transform by calling the `transformDocument` function and passing a GraphQL query to it. 

```ts
import { print } from 'graphql';

const query = gql`
  query TestQuery {
    currentUser {
      name
    }
  }
`;

const documentTransform = new DocumentTransform((document) => {
  // ...
});

const modifiedQuery = documentTransform.transformDocument(query);

console.log(print(modifiedQuery));
// query TestQuery {
//   currentUser {
//     name
//     id
//   }
// }
```

> We use the [`print`](https://graphql.org/graphql-js/language/#print) function exported by `graphql-js` to make the query human-readable.

> Similarly, we can verify that passing a query that _doesn't_ query for `currentUser` is not affected by our transform.

Reviewer 2: ```suggestion
Similarly, we can verify that passing a query that _doesn't_ query for `currentUser` is unaffected by our transform.
```

```ts
const query = gql`
  query TestQuery {
    user {
      name
    }
  }
`;

const modifiedQuery = documentTransform.transformDocument(query);

console.log(print(modifiedQuery));
// query TestQuery {
//   user {
//     name
//   }
// }
```

> ## Composing multiple document transforms

Reviewer 2: ```suggestion
## Composing document transforms
```

> You may have noticed that the `ApolloClient` constructor only takes a single `documentTransform` option. As you add new capabities to your document transforms, you may find that a your document transform function starts to grow unwieldy. The `DocumentTransform` class makes it easy to split your transforms up and compose multiple transforms together into a single transform. 

Reviewer 2: ```suggestion
You may have noticed that the `ApolloClient` constructor only takes a single `documentTransform` option. As you add new capabilities to your document transforms, it may grow unwieldy. The `DocumentTransform` class makes it easy to split your transforms and compose multiple transforms into a single one.
```

Author: I really need to add a spell check to my editor. Thanks!

### Combining multiple document transforms

You can combine multiple document transforms together using the `concat()` function. This forms a "chain" of document transforms that are run one right after the other.

```ts
const documentTransform1 = new DocumentTransform(transform1);
const documentTransform2 = new DocumentTransform(transform2);

const documentTransform = documentTransform1.concat(documentTransform2);
```

> Here `documentTransform1` is combined with `documentTransform2` into a single document transform. Calling the `transformDocument()` function on `documentTransform` will first run the GraphQL document through `documentTransform1`, then through `documentTransform2`. Changes made to the GraphQL document in `documentTransform1` are seen by `documentTransform2`.

Reviewer 2: ```suggestion
Here `documentTransform1` is combined with `documentTransform2` into a single document transform. Calling the `transformDocument()` function on `documentTransform` runs the GraphQL document through `documentTransform1` and then through `documentTransform2`. Changes made to the GraphQL document in `documentTransform1` are seen by `documentTransform2`.
```

#### A note about performance

> Combining multiple transforms is a powerful feature that makes it easy to split up transform logic, which can boost maintainability. It should be noted however that depending on the implementation of your visitor, this can result in the traversal of the GraphQL document AST multiple times. Most of the time this shouldn't be an issue. We recommend using the [`BREAK`](https://graphql.org/graphql-js/language/#break) sentinal from GraphQL where necessary to prevent unnecessary traversal. 

Reviewer 3: ```suggestion
Combining multiple transforms is a powerful feature that makes it easy to split up transform logic, which can boost maintainability. It should be noted however that depending on the implementation of your visitor, this can result in the traversal of the GraphQL document AST multiple times. Most of the time this shouldn't be an issue. We recommend using the [`BREAK`](https://graphql.org/graphql-js/language/#break) sentinel from GraphQL where necessary to prevent unnecessary traversal. 
```

Author: Eagle eyes!! 🦅 Thanks so much for the review!

> Combining multiple transforms is a powerful feature that makes it easy to split up transform logic, which can boost maintainability. It should be noted however that depending on the implementation of your visitor, this can result in the traversal of the GraphQL document AST multiple times. Most of the time this shouldn't be an issue. We recommend using the [`BREAK`](https://graphql.org/graphql-js/language/#break) sentinel from GraphQL where necessary to prevent unnecessary traversal. 

Reviewer 2: ```suggestion
Combining multiple transforms is a powerful feature that makes it easy to split up transform logic, which can boost maintainability. Depending on the implementation of your visitor, this can result in the traversal of the GraphQL document AST multiple times. Most of the time, this shouldn't be an issue. We recommend using the [`BREAK`](https://graphql.org/graphql-js/language/#break) sentinel from GraphQL to prevent unnecessary traversal. 
```

> If, however, you are sending very large queries that require a lot of traversal and have already optimized your visitors with the `BREAK` sentinal, you may find it optimal to combine the transforms into a single visitor that traverses the AST once. 

Reviewer 3: ```suggestion
If, however, you are sending very large queries that require a lot of traversal and have already optimized your visitors with the `BREAK` sentinel, you may find it optimal to combine the transforms into a single visitor that traverses the AST once. 
```

> If, however, you are sending very large queries that require a lot of traversal and have already optimized your visitors with the `BREAK` sentinel, you may find it optimal to combine the transforms into a single visitor that traverses the AST once. 

Reviewer 2: This is a pretty long sentence I found hard to follow.

```suggestion
Suppose you are sending very large queries that require several traversals and have already optimized your visitors with the `BREAK` sentinel. In that case, it's best to combine the transforms into a single visitor that traverses the AST once.
```

See the section on [document caching](#document-caching) to learn how Apollo Client applies optimizations to individual document transforms to mitigate the performance impact when transforming the same GraphQL document multiple times.

### Conditionally running document transforms

At times, you may want to conditionally run a document transform depending on the GraphQL document. You can conditionally run a transform by calling the `split()` static function on the `DocumentTransform` constructor.

```ts
const documentTransform = new DocumentTransform(transform);

const subscriptionTransforms = DocumentTransform.split(

>   (document) => isSubscriptionOperation(document),

Author: Good call! As of 3.8 this will be a function you can export from the library, so I'll just make sure to add the import statement in this code block.

>   (document) => isSubscriptionOperation(document),

Author: Updated this in [9fd4f38](https://github.com/apollographql/apollo-client/pull/11089/commits/9fd4f388fd3f57d6b87a0ce6ffb9945bee6aecfb) to make this clearer. Thanks for the suggestion!

  documentTransform
);
```

> Here the `documentTransform` is only run for subscription operations. For all other operations, no modifications are made to the GraphQL document. Calling the `transformDocument()` function on `subscriptionTransforms` will first check to see if the `document` is a subscription operation, and if so, proceed to run the `documentTransform`. If not, the document transform is bypassed and the GraphQL document is returned as-is.

Reviewer 2: ```suggestion
Here the `documentTransform` is only run for subscription operations. For all other operations, no modifications are made to the GraphQL document. Calling the `transformDocument()` function on `subscriptionTransforms` will first check to see if the `document` is a subscription operation, and if so, proceed to run the `documentTransform`. If not, the document transform is bypassed, and the GraphQL document is returned as-is.
```

The `split` function also allows you to pass a second document transform to its function, allowing you to replicate an if/else condition.

```ts
const subscriptionTransform = new DocumentTransform(transform1);
const defaultTransform = new DocumentTransform(transform2)

const documentTransform = DocumentTransform.split(
  (document) => isSubscriptionOperation(document),
  subscriptionTransform,
  defaultTransform
);
```

Here the `subscriptionTransform` is only run for subscription operations. For all other operations, the GraphQL document is run through the `defaultTransform`.

#### Why should I use the `split()` function instead of a conditional check inside of the transform function?

> It is certainly possible to conditionally run a transform by adding a conditional check inside the transform function itself. 

Reviewer 2: ```suggestion
Sometimes, using the `split()` function is more efficient than running a conditional check inside the transform function.

For example, you can run a transform by adding a conditional check inside the transform function itself:
```

Best to answer the question clearly and then give an example.

```ts
const documentTransform = new DocumentTransform((document) => {
  if (shouldTransform(document)) {
    // ...
    return transformedDocument
  }

  return document
});
```

> Consider the case though where you've combined multiple document transforms together using the `concat()` function.

Reviewer 2: ```suggestion
Consider the case where you've combined multiple document transforms using the `concat()` function:
```

```ts
const documentTransform1 = new DocumentTransform(transform1);
const documentTransform2 = new DocumentTransform(transform2);
const documentTransform3 = new DocumentTransform(transform3);

const documentTransform = documentTransform1
  .concat(documentTransform2)
  .concat(documentTransform3);
```

> The `split()` function makes it easier to skip the _entire_ chain of document transforms.

Reviewer 2: ```suggestion
The `split()` function makes skipping the _entire_ chain of document transforms easier.
```

```ts
const documentTransform = DocumentTransform.split(
  (document) => shouldTransform(document),
  documentTransform1
    .concat(documentTransform2)
    .concat(documentTransform3)
);
```

## Document caching

> You should strive to make your document transforms be deterministic, meaning the document transform should always output the same transformed GraphQL document when given the same input GraphQL document. The `DocumentTransform` class optimizes for this case by caching the transformed result for each input GraphQL document. This speeds up repeated calls to the document transform to avoid unnecessary work.

Reviewer 2: ```suggestion
You should strive to make your document transforms deterministic. This means the document transform should always output the same transformed GraphQL document when given the same input GraphQL document. The `DocumentTransform` class optimizes for this case by caching the transformed result for each input GraphQL document. This speeds up repeated calls to the document transform to avoid unnecessary work.
```

> The `DocumentTransform` class takes this one step further and will record all transformed documents. This means that passing an _already_ transformed document to the `transformDocument()` function will immediately return the GraphQL document.

Reviewer 2: ```suggestion
The `DocumentTransform` class takes this further and records all transformed documents. That means that passing an _already_ transformed document to the `transformDocument()` function will immediately return the GraphQL document.
```

So nice!

```ts
const transformedDocument = documentTransform.transformDocument(document);

transformedDocument === documentTransform.transformDocument(transformedDocument); // => true
```

### Non-deterministic document transforms

> In rare circumstances, you may need to rely on a runtime condition from outside the transform function that changes the result of the document transform. Due to the automatic caching of the document transform, this becomes a problem when that runtime condition changes between calls to `transformDocument()` since your document transform will not know it needs to be run again.

Reviewer 2: ```suggestion
In rare circumstances, you may need to rely on a runtime condition from outside the transform function that changes the result of the document transform. Due to the automatic caching of the document transform, this becomes a problem when that runtime condition changes between calls to `transformDocument()` since your document transform will not know it needs to be rerun.
```

> Instead of completely disabling the document cache in these situations, you can provide a custom cache key that will be used to cache the result of the document transform. This ensures your transform is only called as often as necessary, while maintaining the flexiblity of the runtime condition.

Reviewer 2: ```suggestion
Instead of completely disabling the document cache in these situations, you can provide a custom cache key that will be used to cache the result of the document transform. This ensures your transform is only called as often as necessary while maintaining the flexibility of the runtime condition.
```

To customize the cache key, pass the `getCacheKey` function as an option to the second argument of the `DocumentTransform` constructor. This function receives the `document` that will be passed to your transform function and is expected to return an array.

As an example, here is a document transform that depends on whether the user is connected to the network.

```ts
const documentTransform = new DocumentTransform(
  (document) => {
    if (window.navigator.onLine) {
      // Transform the document when the user is online
    } else {
      // Transform the document when the user is offline
    }
  },
  {
    getCacheKey: (document) => [document, window.navigator.onLine]
  }
);
```

> > It is **HIGHLY** recommended you use the `document` as part of your cache key. In this example, if the `document` is omitted from the cache key, the document transform will only ever output two transformed documents: one for the `true` condition and one for the `false` condition. Using the `document` in the cache key ensures that each unique document in your application will be transformed accordingly.

Reviewer 2: ```suggestion
> ⚠️ **It is highly recommended you use the `document` as part of your cache key.** In this example, if the `document` is omitted from the cache key, the document transform will only output two transformed documents: one for the `true` condition and one for the `false` condition. Using the `document` in the cache key ensures that each unique document in your application will be transformed accordingly.
```

Best to bold the entire important sentence.

Author: Good call!

> You may conditionally disable the cache for select GraphQL documents by returning `undefined` from the `getCacheKey` function. This will force the document transform to run, regardless of whether the input GraphQL document has been seen or not.

Reviewer 2: ```suggestion
You may conditionally disable the cache for select GraphQL documents by returning `undefined` from the `getCacheKey` function. This will force the document transform to run, regardless of whether the input GraphQL document has been seen.
```

```ts
const documentTransform = new DocumentTransform(
  (document) => {
    // ...
  },
  {
    getCacheKey: (document) => {
      // Always run the transform function when `shouldCache` is `false`
      if (shouldCache(document)) {
        return [document] 
      }
    }
  }
);
```

As a last resort, you may completely disable document caching to force your document transform to run each time `transformDocument()` is called. Set the `cache` option to `false` to disable the cache.

```ts
const documentTransform = new DocumentTransform(
  (document) => {
    // ...
  },
  {
    cache: false
  }
);
```

### Caching within combined transforms

> When you combine multiple document transforms together using the `concat()` function, each document transform's cache configuration is honored. This allows you to mix and match transforms that contain varying cache configurations and be confident the transformed GraphQL document is correctly transformed.

Reviewer 2: ```suggestion
When you combine multiple document transforms using the `concat()` function, each document transform's cache configuration is honored. This allows you to mix and match transforms that contain varying cache configurations and be confident the transformed GraphQL document is correctly transformed.
```

```ts
const cachedTransform = new DocumentTransform(transform);

const varyingTransform = new DocumentTransform(transform, {
  getCacheKey: (document) => [document, window.navigator.onLine]
});

const condtionalCachedTransform = new DocumentTransform(transform, {
  getCacheKey: (document) => {
    if (shouldCache(document)) {
      return [document]
    }
  }
});

const nonCachedTransform = new DocumentTransform(transform, {
  cache: false
});

const documentTransform = 
  cachedTransform
    .concat(varyingTransform)
    .concat(conditionalCachedTransform)
    .concat(nonCachedTransform);
```

## TypeScript and GraphQL Code Generator

[GraphQL Code Generator](https://the-guild.dev/graphql/codegen) is a popular tool that generates TypeScript types for your GraphQL documents. It does this by statically analyzing your code to search for GraphQL query strings.

> Document transforms present a challenge for this tool. Because these transforms are used at runtime, we have no way to for static analysis to pick up the changes applied to GraphQL documents from our transform.

Reviewer 2: ```suggestion
Document transforms present a challenge for this tool. Because these transforms are used at runtime, there's no way for static analysis to pick up the changes applied to GraphQL documents from our transform
```

> Thankfully, GraphQL Code Generator provides the ability to run [document transforms](https://the-guild.dev/graphql/codegen/docs/advanced/document-transform) on the documents seen by this tool. We can hook into this ability by using our custom document transform inside the `transform` function passed to the GraphQL Code Generator config.

Reviewer 2: ```suggestion
Thankfully, GraphQL Code Generator can run [document transforms](https://the-guild.dev/graphql/codegen/docs/advanced/document-transform) on the documents seen by this tool. We can hook into this ability by using our custom document transform inside the `transform` function passed to the GraphQL Code Generator config.
```

```ts title="codegen.ts" {2,12-18}
import type { CodegenConfig } from '@graphql-codegen/cli';
import { documentTransform } from './path/to/your/transform';
 
const config: CodegenConfig = {
  schema: 'https://localhost:4000/graphql',
  documents: ['src/**/*.tsx'],
  generates: {
    './src/gql/': {
      preset: 'client',
      documentTransforms: [
        {
          transform: ({ documents }) => {
            return documents.map((documentFile) => {
              documentFile.document = documentTransform.transformDocument(documentFile.document);

              return documentFile;
            });
          }
        }
      ]
    }
  }
}
```

## You might not need document transforms

> Document transforms are a powerful feature of Apollo Client. After reading this article, you may find yourself rushing to find as many use cases for this feature as you can think of. While we encourage you to use this feature where it makes sense in your application, there can be a hidden cost to using it.

Reviewer 2: ```suggestion
Document transforms are a powerful feature of Apollo Client. After reading this article, you may be rushing to find as many use cases for this feature as possible. While we encourage you to use this feature where it makes sense in your application, there can be a hidden cost to using it.
```

> Consider what happens when you're working in a large production application that span many teams within your organization. Document transforms will typically be defined in the code base far away from where GraphQL queries are defined. Not all developers may be aware of their existence nor understand their impact on final GraphQL document.

Reviewer 2: ```suggestion
Consider what happens when working in a large production application that spans many teams within your organization. Document transforms are typically defined in the code base far from where GraphQL queries are defined. Not all developers may be aware of their existence nor understand their impact on final GraphQL document.
```

> Because document transforms have the ability to make endless modifications to GraphQL documents before they are sent to the network, you may find yourself in a position where the result returned from the GraphQL operation does not match the original GraphQL document. This can get especially confusing when document transforms remove fields or make other destructive changes.

Reviewer 2: ```suggestion
Because document transforms can make endless modifications to GraphQL documents before they are sent to the network, you may find yourself in a position where the result returned from the GraphQL operation does not match the original GraphQL document. This can get especially confusing when document transforms remove fields or make other destructive changes.
```

> Consider leaning on existing techniques as a first resort, such as linting. For example, if you require that every selection set in your GraphQL document should include an `id` field, you may find it more useful to create a lint rule that yells at you when you forget to include it. This makes it more obvious exactly what to expect from your GraphQL queries since its located where youre GraphQL queries are defined. Adding an `id` via a document transform makes this relationship an implicit one.

Reviewer 2: Is this a good place to plug https://www.apollographql.com/docs/graphos/delivery/schema-linter/?

Author: Not quite. Schema linter is super useful if you're designing the schema, but Apollo Client is really only concerned with consuming it. [GraphQL ESLint](https://the-guild.dev/graphql/eslint/docs) would be better a better choice in this case. Should I link this tool in some way?

Reviewer 2: Ah, makes sense. I think linking a tool to set folks on the right path makes sense!

> Consider leaning on existing techniques as a first resort, such as linting. For example, if you require that every selection set in your GraphQL document should include an `id` field, you may find it more useful to create a lint rule that yells at you when you forget to include it. This makes it more obvious exactly what to expect from your GraphQL queries since its located where youre GraphQL queries are defined. Adding an `id` via a document transform makes this relationship an implicit one.

Reviewer 2: ```suggestion
Consider leaning on existing techniques as a first resort, such as linting. For example, if you require that every selection set in your GraphQL document should include an `id` field, you may find it more useful to create a lint rule that complains when you forget to include it. This makes it more obvious exactly what to expect from your GraphQL queries since the rule is located where your GraphQL queries are defined. Adding an `id` via a document transform makes this relationship an implicit one.
```

> We encourage you to document your own document transforms to create a shared knowledge base to help avoid confusion. This doesn't mean we consider this feature dangerous. After all, Apollo Client has been performing document transformations for nearly its entire existence and is necessary for its core functionality.

Reviewer 2: ```suggestion
We encourage you to document your own document transforms to create a shared knowledge base to help avoid confusion. This doesn't mean we consider this feature dangerous. After all, Apollo Client has been performing document transformations for nearly its entire existence and they are necessary for its core functionality.
```

### Can I use this to define my own custom directives?

> At a glance, document transforms seem like a great place to create and define custom directives since they can read and modify directives in the GraphQL document. Document transforms however do not have access to the cache, nor can they interact with the data returned from your GraphQL server. If your custom directive needs access to these features in Apollo Client, you will have a difficult time finding ways to make this work.

Reviewer 2: ```suggestion
At a glance, document transforms seem like a great place to create and define custom directives since they can read and modify directives in the GraphQL document. Document transforms, however, don't have access to the cache, nor can they interact with the data returned from your GraphQL server. If your custom directive needs access to these features in Apollo Client, you will have difficulty finding ways to make this work.
```

> As such, custom directives are limited only to those use cases that depend on the GraphQL document itself. 

Reviewer 2: ```suggestion
**Custom directives are limited to use cases that depend on the GraphQL document itself.**
```

This seems to be an important takeaway :)

Author: Good call on the bolded treatment!

const documentTransform3 = new DocumentTransform(transform3);

const documentTransform = documentTransform1
  .concat(documentTransform2)
  .concat(documentTransform3);
```

The `split()` function makes skipping the _entire_ chain of document transforms easier.

```ts
const documentTransform = DocumentTransform.split(
  (document) => shouldTransform(document),
  documentTransform1
    .concat(documentTransform2)
    .concat(documentTransform3)
);
```

## Document caching

You should strive to make your document transforms deterministic. This means the document transform should always output the same transformed GraphQL document when given the same input GraphQL document. The `DocumentTransform` class optimizes for this case by caching the transformed result for each input GraphQL document. This speeds up repeated calls to the document transform to avoid unnecessary work.

The `DocumentTransform` class takes this further and records all transformed documents. That means that passing an _already_ transformed document to the document transform will immediately return the GraphQL document.

```ts
const transformedDocument = documentTransform.transformDocument(document);

> transformedDocument === documentTransform.transformDocument(transformedDocument); // => true

Author: Good call! Fixed in [86b72fa](https://github.com/apollographql/apollo-client/pull/11089/commits/86b72fa00e34968c56e17a3ce17f1993405e9d81)

Document transforms present a challenge for this tool. Because document transforms are used at runtime, there's no way for static analysis to understand the changes applied to GraphQL documents from within document transforms.

Thankfully, GraphQL Code Generator provides a [document transform](https://the-guild.dev/graphql/codegen/docs/advanced/document-transform) feature that allows you to connect Apollo Client document transforms to GraphQL Code Generator. Use your document transform inside the `transform` function passed to the GraphQL Code Generator config:

```ts title="codegen.ts" {2,12-18}
import type { CodegenConfig } from '@graphql-codegen/cli';
import { documentTransform } from './path/to/your/transform';
 
const config: CodegenConfig = {
  schema: 'https://localhost:4000/graphql',
  documents: ['src/**/*.tsx'],
  generates: {
    './src/gql/': {
      preset: 'client',
      documentTransforms: [
        {
          transform: ({ documents }) => {
            return documents.map((documentFile) => {
              documentFile.document = documentTransform.transformDocument(documentFile.document);

              return documentFile;

>             });

Author: Fixed in [86b72fa](https://github.com/apollographql/apollo-client/pull/11089/commits/86b72fa00e34968c56e17a3ce17f1993405e9d81)

          }
        }
      ]
    }
  }
}
```

## You might not need document transforms

Document transforms are a powerful feature of Apollo Client. After reading this article, you may be rushing to find as many use cases for this feature as possible. While we encourage you to use this feature where it makes sense in your application, there can be a hidden cost to using it.

Consider what happens when working in a large production application that spans many teams within your organization. Document transforms are typically defined in the code base far from where GraphQL queries are defined. Not all developers may be aware of their existence nor understand their impact on the final GraphQL document.

Document transforms can make endless modifications to GraphQL documents before they are sent to the network. You may find yourself in a position where the result returned from the GraphQL operation does not match the original GraphQL document. This can get especially confusing when document transforms remove fields or make other destructive changes.

> Consider leaning on existing techniques as a first resort, such as linting. For example, if you require that every selection set in your GraphQL document should include an `id` field, you may find it more useful to create a lint rule that complains when you forget to include it. This makes it more obvious exactly what to expect from your GraphQL queries since the lint rule is applied where your GraphQL queries are defined. Adding an `id` via a document transform makes this relationship an implicit one.

Author: If a transform is used in codegen but not in Apollo Client, wouldn't there potentially be a mismatch in types? Using this example of adding an `id`, if codegen tells you that field is available, but you didn't actually pass that transform to Apollo Client, then `id` wouldn't be requested at runtime.

Forgive me if I'm missing something obvious here. Let me know if you have a good example to provide here to illustrate what you're referring to.

> Consider leaning on existing techniques as a first resort, such as linting. For example, if you require that every selection set in your GraphQL document should include an `id` field, you may find it more useful to create a lint rule that complains when you forget to include it. This makes it more obvious exactly what to expect from your GraphQL queries since the lint rule is applied where your GraphQL queries are defined. Adding an `id` via a document transform makes this relationship an implicit one.

Reviewer 4: The codegen will create your types *and* the `DocumentNode` that you actually use. The `gql` call in your app usually has no direct relationship with the `Document` that is actually used in the end if you use the codegen.
